prov/gni/src/gnix_ep_rcv.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2015 Cray Inc. All rights reserved.
+ * Copyright (c) 2015 Los Alamos National Security, LLC. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * Endpoint common code
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "gnix.h"
+#include "gnix_cm_nic.h"
+#include "gnix_nic.h"
+#include "gnix_util.h"
+#include "gnix_vc.h"
+#include "gnix_ep.h"
+#include "gnix_hashtable.h"
+
+/*******************************************************************************
+ * Helper functions used for handling receipt of messages on GNIX EPs
+ ******************************************************************************/
+
+/* TODO: this function is somewhat of a placeholder till all of the
+ * message pathways are coded. */
+int _gnix_ep_eager_msg_w_data_match(struct gnix_fid_ep *ep, void *msg,
+				   struct gnix_address addr, size_t len,
+				   uint64_t imm, uint64_t sflags)
+{
+	int matched = 0, ret = FI_SUCCESS;
+	struct slist_entry *item, *prev = NULL;
+	struct slist *list;
+	struct gnix_fab_req *req;
+	struct gnix_fid_cq *cq;
+	ssize_t cq_len;
+	uint64_t flags;
+
+	GNIX_TRACE(FI_LOG_EP_DATA, "\n");
+
+	flags = (sflags & FI_REMOTE_CQ_DATA) ? FI_REMOTE_CQ_DATA : 0;
+
+	list = &ep->posted_recv_queue;
+
+	/*
+	 * we have to keep this lock till either we find a match
+	 * or we add the request to the tail of the unexpected queue
+	 */
+
+	fastlock_acquire(&ep->recv_queue_lock);
+
+	for (prev = NULL, item = list->head;
```

Comment:
Can you add a comment for this and others like it for conversion to slist iteration macros when we have them?  If we're not going to keep this code, the never mind.